compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -2978,9 +2995,32 @@ fn add_lld_args(cmd: &mut dyn Linker, sess: &Session, flavor: LinkerFlavor) {
     }
 
     // 1. Implement the "self-contained" part of this feature by adding rustc distribution
-    // directories to the tool's search path:
-    // - if the self-contained linker is enabled on the CLI.
-    if sess.opts.cg.link_self_contained.is_linker_enabled() {
+    // directories to the tool's search path, depending on a mix between what users can specify on
+    // the CLI, and what the target spec enables (as it can't disable components):
+    // - if the self-contained linker is enabled on the CLI or by the target spec,
+    // - and if the self-contained linker is not disabled on the CLI.
+    let self_contained_cli = sess.opts.cg.link_self_contained.is_linker_enabled();
+    let self_contained_target = sess.target.options.link_self_contained.is_linker_enabled();
```

Comment:
This is incorrect, the mingw/musl inference should be accounted for here.

I think `fn self_contained` should return fully inferred `LinkSelfContainedComponents` which should then be passed to this function as a parameter.

Reply:
> I think `fn self_contained` should return fully inferred `LinkSelfContainedComponents` which should then be passed to this function as a parameter.

Sure, I agree this is the goal when the implementation is finalized. However, I don't believe I have the required knowledge about the targets and specific self-contained components features to fully implement this. I'm also not sure how good our test coverage is, so I'm quite worried about introducing regressions while doing that.

I've pushed a commit to build a skeleton of how that could look like, but I'm quite unsure of which components to check for the cases where `self_contained` previously was a bool. It could be considered a WIP commit, and I'm sure there are mistakes, so please take a look.

@rustbot ready

Reply:
Returning `all`/`empty` in the way compatible with the old code should be fine.

Reply:
Yes, this is how the components are emitted, and is done around [here](https://github.com/rust-lang/rust/pull/116035/files#diff-862130c0af71054886913befc8084094485fa7d764d95f55644e033e34acd126R1744-R1747), but then the components need to be read, everywhere where the code used to read the old `self_contained` bool, and turned into e.g. "if the CRTO component is enabled".

All these will succeed because of the compatibility, but the component check themselves may be incomplete, and maybe cause issues if the `all`/`empty` mapping is changed (if e.g. mingw inference starts returning components instead of a bool).